### ISSUE TITLE
shlink-service exited with code 137 when receiving a SIGKILL

### ISSUE THREAD

**Issue by CristianCantoro** [shlinkio/shlink#1524]
#### How Shlink is set-up

* Shlink Version: stable (3.2.1)
* PHP Version: n/a
* How do you serve Shlink: Docker image ``
* Database engine used: PostgreSQL

#### Summary

shlink container exits with error 137 if it receives a SIGKILL

#### Current behavior

shlink container - managed with docker-compose - exits with error 137 if it receives a SIGKILL and does not restart. This is problematic because when docker is updated and restarted the service does not restart gracefully.

#### Expected behavior

The container should handle SIGKILLs gracefully.

#### How to reproduce

Run the container:
```
docker run shlinkio/shlink:stable
```

Kill the container with SIGKILL:
```
docker kill --signal KILL shlink-service
````

The actual setup I am using is available here: [CristianCantoro/my-shlink](https://github.com/CristianCantoro/my-shlink)

**Comment by acelaya** [comment]
Why would it restart if you kill it? Have you passed [--restart](https://docs.docker.com/engine/reference/commandline/run/#restart-policies---restart) to the `docker run` command (or the equivalent option to docker-composer)?

**Comment by CristianCantoro** [comment]
@acelaya wrote:
> Why would it restart if you kill it? Have you passed [--restart](https://docs.docker.com/engine/reference/commandline/run/#restart-policies---restart) to the `docker run` command (or the equivalent option to docker-composer)?

As I said above, the main problem is that when docker is restarted (e.g. when updating it), this service - and only this one - does not restart. The example with SIGKILL was just a way to debug the problem.

And yes, I have passed the `restart` option to docker-compose.

**Comment by acelaya** [comment]
I also use this approach in my personal instance, and I have never experienced any problem.

Can you share your whole docker-compose config, including all the containers, options, env vars, etc? (Except credential).

**Comment by CristianCantoro** [comment]
@acelaya wrote:
> Can you share your whole docker-compose config, including all the containers, options, env vars, etc? (Except credential).

Sure! It was also linked in the comment above. Here it is: [CristianCantoro/my-shlink/docker-compose.yml](https://github.com/CristianCantoro/my-shlink/blob/main/docker-compose.yml)

I am using [wait-for](https://github.com/eficode/wait-for) because otherwise I get this error (the DB is not ready):
```
shlink-service | Creating fresh database if needed...
shlink-web | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
shlink-web | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
shlink-service | 
shlink-web | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
shlink-service | In ExceptionConverter.php line 81:
shlink-service |                                                                                
shlink-service |   An exception occurred in the driver: SQLSTATE[08006] [7] connection to serv  
shlink-service |   er at "db" (172.19.0.2), port 5432 failed: Connection refused                
shlink-service |   	Is the server running on that host and accepting TCP/IP connections?         
shlink-service |                                                                                
shlink-service | 
shlink-web | 10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
shlink-service | In Exception.php line 28:
shlink-service |                                                                                
shlink-service |   SQLSTATE[08006] [7] connection to server at "db" (172.19.0.2), port 5432 fa  
shlink-service |   iled: Connection refused                                                     
shlink-service |   	Is the server running on that host and accepting TCP/IP connections?         
shlink-service |                                                                                
shlink-service | 
shlink-service | In Driver.php line 28:
shlink-service |                                                                                
shlink-service |   SQLSTATE[08006] [7] connection to server at "db" (172.19.0.2), port 5432 fa  
shlink-service |   iled: Connection refused                                                     
shlink-service |   	Is the server running on that host and accepting TCP/IP connections?         
shlink-service |                                                                                
shlink-service | 
shlink-service | db:create
shlink-service | 
```
which means that I need to restart the containers once again to make it work. Using `wait-for` solves the issue. I do not know if you may consider a bug report on the issue above. 

Anyway, from the tests I have done, this does not seem to interfere with the error that I am reporting in this bug. I have tried commenting out the `entrypoint` directive for the `shlink` service in my `docker-compose.yml`, and I get the same behavior as above.

**Comment by acelaya** [comment]
> Using wait-for solves the issue. I do not know if you may consider a bug report on the issue above.

No, that's the expected behavior. Shlink depends on the database so it needs to be up and running before it.

**Comment by acelaya** [comment]
I continue investigating this every time I have some spare time, but haven't found the root cause yet.

I found some useful information here https://gist.github.com/hiroaki-yamamoto/4e27167dd3cb74700ce86e678ff987cc, but still haven't managed to fix it.

Looks like it's a combination of things, including how the entry point is run, how other commands are run inside of it, how processes and signals work, etc.

**Comment by CristianCantoro** [comment]
@acelaya wrote:
> I continue investigating this every time I have some spare time, but haven't found the root cause yet.

Thank you!.

**Comment by mustanggb** [comment]
Some food for thought.

If you're only running one process per container (e.g. openswoole or roadrunner) then you should be able to run your startup script (docker-entrypoint.sh) then `exec` into the aforementioned process running as a daemon, which will switch it to PID 1. Then hopefully the process (openswoole or roadrunner) will correctly handle signals, for itself at least.

However I notice you're not only running one process, i.e. crond is running as a background task. As I understand Docker will only send it's signals to PID 1, so crond will never see it. I don't know how crond handles this, but I'm wondering if this could be the unknown source of the exit code?

Perhaps you could try running a supervisor/init as PID 1, then openswoole/roadrunner and crond as it's two services.

My suggestion for such task is [s6-overlay](https://github.com/just-containers/s6-overlay).
